StRoot/StFwdTrackMaker/StFwdTrackMaker.cxx

```diff
@@ -737,22 +742,113 @@ void StFwdTrackMaker::loadFttHitsFromGEANT( FwdDataSource::McTrackMap_t &mcTrack
+ * 
+ * Order of precedence:
+ * MuDst StMuFstCollection (Data)
+ * StEvent StFstHitCollection (Data or slowsim)
+ * StEvent StRndHitCollection (fast sim)
+ * GEANT St_g2t_fts_hit (starsim only) - note if rasterizer is active this takes priority over FastSim
+ * 
+ * @param mcTrackMap : MC track map if running sim
+ * @param hitMap : FST hitmap to populate
+ * @param count  : number of hits loaded
+ */
+int StFwdTrackMaker::loadFstHits( FwdDataSource::McTrackMap_t &mcTrackMap, FwdDataSource::HitMap_t &hitMap ){
+    
+    int count = loadFstHitsFromMuDst(mcTrackMap, hitMap); 
+    if ( count > 0 ) return count; // only load from one source at a time
+    
+    count += loadFstHitsFromStEvent(mcTrackMap, hitMap); 
+    if ( count > 0 ) return count; // only load from one source at a time
+    
+    bool siRasterizer = mFwdConfig.get<bool>( "SiRasterizer:active", false );
+
+    if ( !siRasterizer ) count += loadFstHitsFromStEventFastSim( mcTrackMap, hitMap );
+    if ( count > 0 ) return count; // only load from one source at a time
+
+    return loadFstHitsFromGEANT( mcTrackMap, hitMap );    
+} // loadFstHits
+
+int StFwdTrackMaker::loadFstHitsFromMuDst( FwdDataSource::McTrackMap_t &mcTrackMap, FwdDataSource::HitMap_t &hitMap){
+    int count = 0;
+    StMuDstMaker *mMuDstMaker = (StMuDstMaker *)GetMaker("MuDst");
+    if(!mMuDstMaker) {
+        LOG_WARN << " No MuDstMaker ... bye-bye" << endm;
+        return 0;
+    }
+    StMuDst *mMuDst = mMuDstMaker->muDst();
+    if(!mMuDst) {
+        LOG_WARN << " No MuDst ... bye-bye" << endm;
+        return 0;
+    }
+    
+    StMuFstCollection * fst = mMuDst->muFstCollection();
+    if (!fst) {
+        LOG_WARN << "No StMuFstCollection ... bye-bye" << endm;
+        return 0;
+    }
+
+    LOG_INFO << "Loading " << fst->numberOfHits() << " StMuFstHits" << endm;
+    TMatrixDSym hitCov3(3);
+    for ( unsigned int index = 0; index < fst->numberOfHits(); index++){
+        StMuFstHit * muFstHit = fst->getHit( index );
+
+        float vR = muFstHit->localPosition(0);
+        float vPhi = muFstHit->localPosition(1);
+        float vZ = muFstHit->localPosition(2);
+
+        const float dz0 = fabs( vZ - 151.75 );
+        const float dz1 = fabs( vZ - 165.248 );
+        const float dz2 = fabs( vZ - 178.781 );
+
+        int d = 0 * ( dz0 < 1.0 ) + 1 * ( dz1 < 1.0 ) + 2 * ( dz2 < 1.0 );
+
+        float x0 = vR * cos( vPhi );
+        float y0 = vR * sin( vPhi );
+        hitCov3 = makeSiCovMat( TVector3( x0, y0, vZ ), mFwdConfig );
+
+        LOG_DEBUG << "FST HIT: d = " << d << ", x=" << x0 << ", y=" << y0 << ", z=" << vZ << endm;                
+        mFstHits.push_back( TVector3( x0, y0, vZ)  );
+
+        // we use d+4 so that both FTT and FST start at 4
+        FwdHit *hit = new FwdHit(count++, x0, y0, vZ, d+4, 0, hitCov3, nullptr);
+        // Add the hit to the hit map
+        hitMap[d+4].push_back(hit);
```

Comment:
I think this is the case where you want to be a bit more explicit with the code. You may know that you use smart pointers but it does look scary to a casual reviewer who is not aware of the internals.

```suggestion
        auto hit = std::make_unique<FwdHit>(count++, hitInfo.x, hitInfo.y, hitInfo.z, hitInfo.d + START_INDEX_FOR_FST, 0, hitCov3, nullptr);
        hitMap[d + START_INDEX_FOR_FST].push_back(std::move(hit));
```

And BTW, why not create the hits on the stack?